src/QMCHamiltonians/MomentumEstimator.cpp

```diff
@@ -172,115 +147,271 @@ void MomentumEstimator::setParticlePropertyList(PropertySetType& plist
   if (!hdf5_out)
   {
     copy(nofK.begin(),nofK.end(),plist.begin()+myIndex+offset);
-    copy(compQ.begin(),compQ.end(),plist.begin()+myIndex+nofK.size()+offset);
   }
 }
 
 bool MomentumEstimator::putSpecial(xmlNodePtr cur, ParticleSet& elns, bool rootNode)
 {
   OhmmsAttributeSet pAttrib;
   std::string hdf5_flag="yes";
+  kmax=0.0;
+  kmax0=0.0;
+  kmax1=0.0;
+  kmax2=0.0;
   pAttrib.add(hdf5_flag,"hdf5");
-  pAttrib.add(kgrid,"grid");
-  pAttrib.add(M,"samples");
+  //pAttrib.add(kgrid,"grid");
+  pAttrib.add(kmax,"kmax");
+  pAttrib.add(kmax0,"kmax0");
+  pAttrib.add(kmax1,"kmax1");
+  pAttrib.add(kmax2,"kmax2");
+  pAttrib.add(M,"samples"); // default value is 40 (in the constructor)
   pAttrib.put(cur);
   hdf5_out = (hdf5_flag=="yes");
+  RealType min_Length=1.0e10;
+  PosType vec_length(0);
+  for (int i=0; i<OHMMS_DIM; i++)
+  {
+    PosType a(0);
+    for (int j=0; j<OHMMS_DIM; j++)
+      a[j]=elns.Lattice.G(j,i);
+    //length of reciprocal lattice vector
+    vec_length[i]=2.0*M_PI*std::sqrt(dot(a,a));
+    if (min_Length>vec_length[i])
+      min_Length=vec_length[i];
+  }
 #if OHMMS_DIM==3
-  int numqtwists(6*kgrid+3);
-  std::vector<int> qk(0);
-  mappedQtonofK.resize(numqtwists,qk);
-  compQ.resize(numqtwists);
-  RealType qn(4.0*M_PI*M_PI*std::pow(Lattice.Volume,-2.0/3.0));
-  mappedQnorms.resize(numqtwists,qn*0.5/RealType(M));
-  if (twist[0]==0)
-    mappedQnorms[kgrid]=qn/RealType(M);
-  if (twist[1]==0)
-    mappedQnorms[3*kgrid+1]=qn/RealType(M);
-  if (twist[2]==0)
-    mappedQnorms[5*kgrid+2]=qn/RealType(M);
-  Q.resize(numqtwists);
-  for (int i=-kgrid; i<(kgrid+1); i++)
+  PosType kmaxs(0);
+  kmaxs[0]=kmax0;
+  kmaxs[1]=kmax1;
+  kmaxs[2]=kmax2;
+  bool sphere_only=true;
+  bool directional_only=true;
```

Comment:
All the XXX_only logics are messy. _only implicitly impose the state of the other. The I saw later !sphere_only was doing direction. !direction_only is doing sphere.

Reply:
Attempt here is to do one of three things: 1) k-points in a sphere, 2) k-points in a parallelepiped, 3) both at the same time (better than making a second estimator).

Fusing into a single loop+logic is somewhat ugly but obviates the need to track (e.g. in a set/hash table) which k-points have been selected already.

Reply:
Agree that naming+logic can be improved.

Reply:
I don't mean fusing into 1 logic. The problem is ! XXX_only double negative. I think you can directly use sphere and directional instead of the opposite meaning.